### ISSUE TITLE
Data model doesn't exactly fit GeoSPARQL standard

### ISSUE THREAD

**Issue by LorenzBuehmann** [ad-freiburg/osm2rdf#24]
Hi,

used the converter and according to what I saw in the output, the generated data doesn't match the GeoSPARQL standard, see https://opengeospatial.github.io/ogc-geosparql/geosparql11/spec.html#_core

In particular I mean that `geo:hasGeometry` should only point to a geometry and not to the serialization directly. I know it might be annoying and it will introduce an additional triple per geometry, but this is just what the standard says.

Without this I guess some of the standard triple stores won't be able to make use of the geospatial data, e.g. for indexing I think.

If not already done, maybe you could mention this somewhere in your documentation at least?

Cheers,
Lorenz

**Comment by hannahbast** [comment]
@LorenzBuehmann Thanks a lot for paying attention. Since all the queries that we are currently interested in, want the literal, this is a bit annoying. I wonder what the best solution is. Here are some options I see:

1. Deviate from the standard and document it (as you suggest). This sounds unreasonable at first, but even the Wikidata Query Service deviates from Wikidata's own RDF serialization for very similar reasons. For example, according to their RDF dump, you need `^schema:about/wikibase:sitelinks` to get from an entity to its number of sitelinks, whereas in the WDQS only `wikibase:sitelinks` works. I guess their reason is mostly to avoid the very expensive join and as a side-effect, it's also more convenient.

2. Let the QLever UI suggest `geo:hasGeometry/geo:hasSerialization` in lieu of `geo:hasGeometry` and use the precomputed join when the former appears in a query. This would be easy to do. It would be a bit annoying because it would break all example queries floating around.

3. Introduce our own predicate, for example `osm2rdf:hasGeometry`, for the shortcut. But that would also break all example queries floating around. 

Opinions?

**Comment by LorenzBuehmann** [comment]
Yep, I know - I can see why they work on Geometry -> Serialization abstraction, but in reality it's quite often just more effort for users (and triple stores)

1. would indeed be the easiest way
2. could be more specific with `geo:hasGeometry/geo:asWKT` to denote the kind of serialization - for querying indeed internally rewriting to a simple lookup of `geo:hasGeometry` would work - but that is QLever specific then - I'm more concerned about interoperability of the RDF dataset itself - other triple stores won't have this rewriting feature - currently, I'm doing an SPARQL Update statement to make the data working with other triple stores
3. would at least not violate the GeoSPARQL standard, but then people would ask "why can't you just deliver the data in the standard, why not reusing an existing vocabulary?"

Decisions are hard here. I guess option 2. would make running standard GeoSPARQL queries work in the UI to some extent - I mean, so far you don't support any standard function, but materialize the geospatial relations in advance (which is nice of course)

One thing I definitely would support is to at least add an option in the converter that allows for generating the data as `geo:hasGeometry/geo:asWKT` - this would avoid any otherwise necessary post-processing

By the way, you have a lots of options as far as I can see from code, but they're hidden exclusively in the C++ code? There isn't a config file or the like, right?

**Comment by l00mi** [comment]
I was trying to create an Spatial Index with Fuseki on some _osm2rdf_ files today, and only after quite some time found this issue.

The error fore the spatial-index code throws in this case is the following (for reference)
```
xception in thread "main" org.apache.jena.rdf.model.ResourceRequiredException: "POINT(50.6691628 26.2873715)"^^http://www.opengis.net/ont/geosparql#wktLiteral
        at org.apache.jena.rdf.model.impl.StatementBase.mustBeResource(StatementBase.java:144)
        at org.apache.jena.rdf.model.impl.StatementImpl.getResource(StatementImpl.java:92)
        at org.apache.jena.geosparql.spatial.SpatialIndex.getGeometryLiteralIndexItems(SpatialIndex.java:406)
        at org.apache.jena.geosparql.spatial.SpatialIndex.getSpatialIndexItems(SpatialIndex.java:381)
        at org.apache.jena.geosparql.spatial.SpatialIndex.findSpatialIndexItems(SpatialIndex.java:299)
        at org.apache.jena.geosparql.spatial.SpatialIndex.buildSpatialIndex(SpatialIndex.java:240)
        at org.apache.jena.geosparql.configuration.GeoSPARQLConfig.setupSpatialIndex(GeoSPARQLConfig.java:313)
        at com.zazuko.spatialindexer.SpatialIndexer.setupSpatialIndexWithoutSrsUri(SpatialIndexer.java:34)
        at com.zazuko.spatialindexer.SpatialIndexer.main(SpatialIndexer.java:104)
```
I would heavily advice to comply to the GeoSPARQL ontology here, to be able to use other tools complying the ontology. ( Somehow one of the main ideas of an ontology, no? (-: )

Meanwhile the best would be to make a note on the README about this non-compliance.

**Comment by patrickbr** [comment]
@l00mi, I agree. We are currently working on this and will add an option to osm2rdf to store geometries as `geo:hasGeometry/geo:asWKT`, as suggested by @LorenzBuehmann above.

**Comment by l00mi** [comment]
Very nice, thank you. As I am planning a project based on your work, might I ask if you have a very rough ETA for this? If not its okay ... I have a work around for now.

**Comment by patrickbr** [comment]
I will create a PR for this today.

**Comment by patrickbr** [comment]
c091889 adds an option `--add-geometry-objects` which writes intermediate geometry objects for `geo:hasGeometry`. These then point via `geo:asWKT` to their WKT literal. See also PR  #48. Feel free to test this :)

**Comment by LorenzBuehmann** [comment]
Nice, I had a custom fork where I did the cide changes by myself - but always better to keep to the original project. Thanks.

**Comment by l00mi** [comment]
Yes also from me, thank you. I will give it a try tomorrow. 

A proposition: I would go the other way around, make your PR the standard, and provide a flag in the realms of --legacy-geometry-export ? WDYT?

**Comment by patrickbr** [comment]
@l00mi, yes, @lehmann-4178656ch and I discussed this yesterday. We will adjust the parameter in the PR, making the changes the standard.

**Comment by l00mi** [comment]
Very good, looking forward that this lands to try out.

**Comment by lehmann-4178656ch** [comment]
> Can you roughly explain what kind of URI pattern you use for the geometries? Just asking because I did and normally stick to append as simple `/geometry` (if possible) - at least I try to avoid bnodes as those are still horrible for some tools.

We decided to not pollute the OSM namespace further with this addition and we are adding them as explicit objects in the osm2rdfgeom namespace which currently translates to `https://osm2rdf.cs.uni-freiburg.de/rdf/geom#`

I've just added some tests to showing both output possibilities e.g.:
```
osmway:21 geo:hasGeometry "MULTIPOLYGON(((48.0 7.5,48.0 7.6,48.1 7.6,48.1 7.5,48.0 7.5)))"^^geo:wktLiteral .
```
```
osmway:21 geo:hasGeometry osm2rdfgeom:wayarea_21 .
osm2rdfgeom:wayarea_21 geo:asWKT MULTIPOLYGON(((48.0 7.5,48.0 7.6,48.1 7.6,48.1 7.5,48.0 7.5)))^^geo:wktLiteral .
```

**Comment by l00mi** [comment]
Thank you for solving this. I can confirm that I was just able to create a [Spatial Index](https://jena.apache.org/documentation/geosparql/geosparql-fuseki.html#fuseki-assembler) with the latest osm2rdf conversion.

**Comment by patrickbr** [comment]
Great, thank you for your feedback.
